src/memtest.rs

```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+/// memtest86plus/memtest86plus?tab=readme-ov-file#memtest86-test-algorithms)
+///
+/// This test runs the moving inversion algorithm with a random pattern for every memory location.
+#[tracing::instrument(skip_all)]
+pub fn test_mov_inv_random<O: TestObserver>(
+    memory: &mut [usize],
+    mut observer: O,
+) -> MemtestResult<O> {
+    use {
+        rand::{rngs::SmallRng, Rng, SeedableRng},
+        std::time::Instant,
+    };
+    let expected_iter = u64::try_from(memory.len())
+        .ok()
+        .and_then(|count| count.checked_mul(MOV_INV_ITERATIONS * 2))
+        .context("Total number of iterations overflowed")?;
+    observer.init(expected_iter);
+    let seed = {
+        let time_bytes = Instant::now().elapsed().as_nanos().to_le_bytes();
+        let mut seed = [0; 32];
+        seed[0..16].copy_from_slice(&time_bytes);
+        seed[16..32].copy_from_slice(&time_bytes);
+        seed
+    };
+
+    let mut rng = SmallRng::from_seed(seed);
+    for mem_ref in memory.iter_mut() {
+        observer.check().map_err(|e| MemtestError::Observer(e))?;
+        write_volatile_safe(mem_ref, rng.gen());
+    }
+
+    let mut rng = SmallRng::from_seed(seed);
+    for mem_ref in memory.iter_mut() {
+        observer.check().map_err(|e| MemtestError::Observer(e))?;
+        let address = address_from_ref(mem_ref);
+        let expected = rng.gen();
+        let actual = read_volatile_safe(mem_ref);
+
+        if actual != expected {
+            info!("Test failed at 0x{address:x}");
+            return Ok(MemtestOutcome::Fail(MemtestFailure::UnexpectedValue {
+                address,
+                expected,
+                actual,
+            }));
+        }
+
+        write_volatile_safe(mem_ref, !expected);
+    }
+
+    let mut rng = SmallRng::from_seed(seed);
+    for mem_ref in memory.iter_mut() {
+        observer.check().map_err(|e| MemtestError::Observer(e))?;
+        let address = address_from_ref(mem_ref);
+        let expected = !rng.gen::<usize>();
+        let actual = read_volatile_safe(mem_ref);
+
+        if actual != expected {
+            info!("Test failed at 0x{address:x}");
+            return Ok(MemtestOutcome::Fail(MemtestFailure::UnexpectedValue {
+                address,
+                expected,
+                actual,
+            }));
+        }
+
+        write_volatile_safe(mem_ref, !expected);
+    }
+
+    Ok(MemtestOutcome::Pass)
+}
+
+/// This test uses the Modulo-20 algorithm implemented by [memtest86+](https://github.com/
+/// memtest86plus/memtest86plus),  which is designed to avoid effects of caching and buffering.
+///
+/// The test generates a random value, then write the value to every 20th memory location.
+/// Afterwards write the complement of the value to all other locations. Then verify that the values
+/// stored in every 20th location is unchanged.
+/// The procedure is repeated with offsets 0-19 to test all memory locations.
```

Comment:
```suggestion
/// The test generates a random value, then write the value to every 20th memory location.
/// Afterwards write the complement of the value to all other locations one or more times (twice in
/// this case). Then verify that the values stored in every 20th location is unchanged.
///
/// The procedure is repeated with offsets 0-19 to test all memory locations.
```
Small correction to comments